### PR TITLE
fix: Form `AutovalidateMode.onUserInteraction` behavior

### LINKED ISSUE #107350: [Form] `AutovalidateMode.onUserInteraction` validates Textfields without interaction
## Steps to Reproduce

Execute `flutter run` on the code sample 

**Expected results:** 

https://user-images.githubusercontent.com/85326522/178113909-a6705448-d7c1-4d65-acdd-6b28342b0fcf.mov

**Actual results:** 

https://user-images.githubusercontent.com/85326522/178113919-f6e130ec-dcc0-4bd7-9303-dd8f93ec60d1.mov

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final _formKey = GlobalKey<FormState>();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: SingleChildScrollView(
          padding: const EdgeInsets.all(20),
          child: Column(
            children: [
              const Text('has issue onUserInteraction (wraped with Form)'),
              Form(
                key: _formKey,
                autovalidateMode: AutovalidateMode.onUserInteraction,
                child: Column(
                  children: [
                    TextFormField(
                      decoration: const InputDecoration(
                        hintText: 'Enter your Name',
                      ),
                      validator: (val) {
                        return val!.length < 4 ? 'Not valid' : null;
                      },
                    ),
                    const SizedBox(height: 20),
                    TextFormField(
                      decoration: const InputDecoration(
                        hintText: 'Enter your email Id',
                      ),
                      validator: (val) {
                        return val!.length < 4 ? 'email not valid' : null;
                      },
                    ),
                    const SizedBox(height: 20),
                    TextFormField(
                      decoration: const InputDecoration(
                        hintText: 'Enter your password',
                      ),
                      obscureText: true,
                      validator: (val) {
                        return val!.length < 4 ? 'password not valid' : null;
                      },
                    ),
                  ],
                ),
              ),
              const SizedBox(height: 40),
              const Text('Working fine onUserInteraction (wraped without Form)'),
              TextFormField(
                decoration: const InputDecoration(
                  hintText: 'Enter your Name',
                ),
                autovalidateMode: AutovalidateMode.onUserInteraction,
                validator: (val) {
                  return val!.length < 4 ? 'Not valid' : null;
                },
              ),
              const SizedBox(height: 20),
              TextFormField(
                decoration: const InputDecoration(
                  hintText: 'Enter your email Id',
                ),
                autovalidateMode: AutovalidateMode.onUserInteraction,
                validator: (val) {
                  return val!.length < 4 ? 'email not valid' : null;
                },
              ),
              const SizedBox(height: 20),
              TextFormField(
                decoration: const InputDecoration(
                  hintText: 'Enter your password',
                ),
                autovalidateMode: AutovalidateMode.onUserInteraction,
                obscureText: true,
                validator: (val) {
                  return val!.length < 4 ? 'password not valid' : null;
                },
              ),
            ],
          ),
        ),
      ),
    );
  }
}


```

</details>

<details>
  <summary>Logs</summary>

```
[✓] Flutter (Channel stable, 3.0.2, on macOS
    12.4 21F79 darwin-arm, locale en-IN)
    • Flutter version 3.0.2 at
      /Users/newton/Development/flutter
    • Upstream repository
      https://github.com/flutter/flutter.git
    • Framework revision cd41fdd495 (4 weeks
      ago), 2022-06-08 09:52:13 -0700
    • Engine revision f15f824b57
    • Dart version 2.17.3
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android
    devices (Android SDK version 32.0.0-rc1)
    • Android SDK at
      /Users/newton/Library/Android/sdk
    • Platform android-32, build-tools
      32.0.0-rc1
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/Contents/Home/bi
      n/java
    • Java version OpenJDK Runtime Environment
      (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode
    13.4.1)
    • Xcode at
      /Applications/Xcode.app/Contents/Develop
      er
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google
      Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨
      https://plugins.jetbrains.com/plugin/921
      2-flutter
    • Dart plugin can be installed from:
      🔨
      https://plugins.jetbrains.com/plugin/635
      1-dart
    • Java version OpenJDK Runtime Environment
      (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.68.1)
    • VS Code at /Applications/Visual Studio
      Code.app/Contents
    • Flutter extension version 3.42.0

[✓] Connected device (3 available)
    • sdk gphone64 arm64 (mobile) •
      emulator-5554 • android-arm64  • Android
      12 (API 31) (emulator)
    • macOS (desktop)             • macos
      • darwin-arm64   • macOS 12.4 21F79
      darwin-arm
    • Chrome (web)                • chrome
      • web-javascript • Google Chrome
      103.0.5060.53

[✓] HTTP Host Availability
    • All required HTTP hosts are available

```

</details>

### PR DESCRIPTION
**Pull request by pedromassango**
*List which issues are fixed by this PR. You must list at least one issue.*
Fixes https://github.com/flutter/flutter/issues/107350

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/form.dart [hunk 1]
```diff
@@ -237,8 +237,17 @@ class FormState extends State<Form> {
 
   bool _validate() {
     bool hasError = false;
+    bool validateFormField(FormFieldState<dynamic> field) =>
+        !field.validate() || hasError;
+
     for (final FormFieldState<dynamic> field in _fields) {
-      hasError = !field.validate() || hasError;
+      if (widget.autovalidateMode == AutovalidateMode.onUserInteraction) {
+        if (_hasInteractedByUser && field._hasInteractedByUser.value) {
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/form.dart:245]: Is there a documentation on what Form.autovalidateMode will interact with FormField.autovalidateMode if they are both set? If not, we should document it.

Comment (chunhtai) [inline packages/flutter/lib/src/widgets/form.dart:245]: also, should we remove _hasInteractedByUser here? It seems like _hasInteractedByUser will always be true if field._hasInteractedByUser.value is true..

**Comment by Hixie** [comment]
@pedromassango Is this PR still on your radar?

**Comment by pedromassango** [comment]
> @pedromassango Is this PR still on your radar?

Yes

File: packages/flutter/lib/src/widgets/form.dart [hunk 2]
```diff
@@ -237,8 +237,17 @@ class FormState extends State<Form> {
...
+          hasError = validateFormField(field);
+        }
+      } else {
+        hasError = validateFormField(field);
+      }
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/form.dart:250]: Nit: If I have the logic correct, maybe this is cleaner?

```suggestion
      if (widget.autovalidateMode != AutovalidateMode.onUserInteraction
          || (_hasInteractedByUser && field._hasInteractedByUser.value)) {
        hasError = validateFormField(field);
      }
```

**Review by justinmc — COMMENTED** [review]
Related to: https://github.com/flutter/flutter/issues/125766

CC @Renzo-Olivares who's looking into that bug.

The PR looks good to me if we're on board with this behavior.

**Comment by Hixie** [comment]
@pedromassango Hey sorry to badger you, just wondering if we should close this one for now to get it off the review queue or if it's something you expect to get back to in the near future. Obviously no rush either way.

**Comment by Hixie** [comment]
@pedromassango I'm going to close this for now, please don't hesitate to reopen it if this is something you'd like to continue working on. Thanks!

**Pull request closed without merging**